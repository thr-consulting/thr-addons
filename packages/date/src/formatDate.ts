import {DateTimeFormatter, LocalDate, LocalDateTime, LocalTime, ZonedDateTime, ZoneId} from '@js-joda/core';
// @ts-ignore
import {Locale} from '@js-joda/locale_en-us';
import {toLocalDate, toLocalDateTime, ILocalDateLike, isLocalDateLike} from './conversion';

export enum FormatDateType {
	short = 'short',
	medium = 'medium',
	long = 'long',
}

export interface FormatDateParams {
	type?: FormatDateType | 'short' | 'medium' | 'long';
	time?: boolean;
	date?: boolean;
	format?: string;
}

/**
 * Formats a date to a string
 * @param obj
 * @param params
 * @param zone
 */
export function formatDate(
	obj: ILocalDateLike | Date | LocalDate | number | string | LocalDateTime | LocalTime | ZonedDateTime | null | undefined,
	{type = FormatDateType.short, time = false, date = true, format}: FormatDateParams = {type: FormatDateType.short, time: false, date: true},
	zone: ZoneId = ZoneId.SYSTEM,
) {
	let l: LocalDate | LocalDateTime | LocalTime | ZonedDateTime | undefined;
	if (!obj) return '';
	if (typeof obj === 'number') l = toLocalDate(obj);
	if (typeof obj === 'string' && time === false) l = toLocalDate(obj);
	if (typeof obj === 'string') l = toLocalDateTime(obj);
	if (isLocalDateLike(obj)) l = toLocalDate(obj);
	if (obj instanceof Date) l = toLocalDateTime(obj, zone);
	if (obj instanceof LocalDate) l = obj;
	if (obj instanceof LocalDateTime) l = obj;
	if (obj instanceof LocalTime) l = obj;
	if (obj instanceof ZonedDateTime) l = obj;
	if (!l) throw new Error('Date could not be converted to a js-joda date');

	let dateFormat: string;
	let timeFormat: string;
	switch (type) {
		case FormatDateType.medium:
			dateFormat = 'MMM d, yyyy';
			timeFormat = 'h:mm a';
			break;
		case FormatDateType.long:
			dateFormat = 'MMMM d, yyyy';
			timeFormat = 'h:mm a';
			break;
		case FormatDateType.short:
		default:
			dateFormat = 'M/d/yyyy';
			timeFormat = 'h:mm a';
			break;
	}

	if (l instanceof LocalDate) timeFormat = '';
	if (l instanceof LocalTime) dateFormat = '';

	let formatString = !date ? '' : dateFormat;
	if (time) {
		formatString = `${formatString} ${timeFormat}`.trim();
	}

	const formatter = DateTimeFormatter.ofPattern(format || formatString).withLocale(Locale.ENGLISH);

	return l.format(formatter).replace(/AM|PM/, x => x.toLowerCase());
}
