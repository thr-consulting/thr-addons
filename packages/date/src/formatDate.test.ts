import {Instant, LocalDate, LocalDateTime, ZonedDateTime, ZoneId, ZoneOffset, LocalTime} from '@js-joda/core';
import {formatDate, FormatDateType} from './formatDate';

expect.addSnapshotSerializer({
	test: v => v instanceof Date,
	print: (v: any) => v.toJSON(),
});

expect.addSnapshotSerializer({
	test: v => v instanceof LocalDate || v instanceof LocalDateTime || v instanceof ZonedDateTime,
	print: (v: any) => v.toJSON(),
});

function getDateVars() {
	return {
		// Actual UTC Epoch
		epochDate: new Date('1970-01-01T00:00:00.000Z'),
		epochLocalDateTime: LocalDateTime.ofEpochSecond(0, ZoneOffset.UTC),
		epochZonedDateTime: ZonedDateTime.ofInstant(Instant.EPOCH, ZoneId.UTC),
		epochString: '1970-01-01T00:00:00.000Z',

		// Local time epoch (depends on TZ environment variable)
		localEpochDate: new Date('1970-01-01T00:00:00.000'),
		localEpochLocalDate: LocalDate.ofEpochDay(0),
		localEpochDays: 0,
		localEpochObject: {year: 1970, month: 1, day: 1},
		localEpochLocalDateTime: LocalDateTime.of(1970, 1, 1, 0, 0, 0, 0),
		localEpochStringA: '1970-01-01',
		localEpochStringB: '1970-01-01T00:00:00.000',

		// Local time epoch + 23 hours
		e23Date: new Date('1970-01-01T23:00:00.000'),
		e23LocalDateTime: LocalDateTime.of(1970, 1, 1, 23, 0, 0, 0),
		e23String: '1970-01-01T23:00:00.000',
	};
}

describe('Formatting', () => {
	it('should format with defaults', () => {
		const vars = getDateVars();
		expect(formatDate(vars.epochDate, undefined, ZoneId.UTC)).toMatchSnapshot();
		expect(formatDate(vars.localEpochDate, undefined, ZoneId.SYSTEM)).toMatchSnapshot();
	});
	it('should format with default string', () => {
		const vars = getDateVars();
		expect(formatDate(vars.epochString, {time: true}, ZoneId.UTC)).toMatchSnapshot();
		expect(formatDate(vars.localEpochStringB, {time: true}, ZoneId.SYSTEM)).toMatchSnapshot();
		expect(formatDate(vars.epochString, undefined, ZoneId.UTC)).toMatchSnapshot();
		expect(formatDate(vars.localEpochStringB, undefined, ZoneId.SYSTEM)).toMatchSnapshot();
	});
	it('should format with options', () => {
		const vars = getDateVars();
		expect(formatDate(vars.localEpochLocalDateTime, {type: FormatDateType.short, time: true, date: true})).toMatchSnapshot();
		expect(formatDate(vars.localEpochLocalDateTime, {type: FormatDateType.medium, time: true, date: true})).toMatchSnapshot();
		expect(formatDate(vars.localEpochLocalDateTime, {type: FormatDateType.long, time: true, date: true})).toMatchSnapshot();
	});
	it('should format with individual options', () => {
		const vars = getDateVars();
		expect(formatDate(vars.localEpochLocalDateTime, {type: FormatDateType.short})).toMatchSnapshot();
		expect(formatDate(vars.localEpochLocalDateTime, {type: FormatDateType.medium})).toMatchSnapshot();
		expect(formatDate(vars.localEpochLocalDateTime, {type: FormatDateType.long})).toMatchSnapshot();

		expect(formatDate(vars.localEpochLocalDateTime, {time: true})).toMatchSnapshot();
		expect(formatDate(vars.localEpochLocalDateTime, {time: false})).toMatchSnapshot();

		expect(formatDate(vars.localEpochLocalDateTime, {date: true})).toMatchSnapshot();
		expect(formatDate(vars.localEpochLocalDateTime, {date: false, time: true})).toMatchSnapshot();
		expect(formatDate(vars.localEpochLocalDateTime, {date: false, time: false})).toMatchSnapshot();
	});
	it('should use a custom date format', () => {
		const vars = getDateVars();
		expect(formatDate(vars.localEpochLocalDateTime, {format: 'MMMM_d_yyyy_h_mm_a'})).toMatchSnapshot();
	});
	it('should format an integer (epoch days)', () => {
		expect(formatDate(17400)).toMatchSnapshot();
	});
	it('should format a LocalDate', () => {
		expect(formatDate(LocalDate.ofEpochDay(17400))).toMatchSnapshot();
	});
	it('should format a JS Date', () => {
		expect(formatDate(new Date(2017, 5, 23, 6, 0, 0))).toMatchSnapshot();
	});
	it('should format a LocalDateTime', () => {
		expect(formatDate(LocalDateTime.ofEpochSecond(1740, ZoneOffset.UTC))).toMatchSnapshot();
	});

	it('should format a falsy value', () => {
		expect(formatDate(null)).toEqual('');
	});

	it('should format a LocalTime', () => {
		expect(formatDate(LocalTime.of(0, 0, 0, 0), {time: true, date: false})).toMatchSnapshot();
	});

	// it('should throw an error when a bad value is passed in', () => {
	// 	// @ts-ignore
	// 	expect(formatDate('A')).toThrowErrorMatchingSnapshot(); // .toThrowError('Date could not be converted to a js-joda date');
	// });
});
