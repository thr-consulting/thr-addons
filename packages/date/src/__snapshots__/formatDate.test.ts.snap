// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Formatting should format a JS Date 1`] = `"6/23/2017"`;

exports[`Formatting should format a LocalDate 1`] = `"8/22/2017"`;

exports[`Formatting should format a LocalDateTime 1`] = `"1/1/1970"`;

exports[`Formatting should format a LocalTime 1`] = `"12:00 am"`;

exports[`Formatting should format an integer (epoch days) 1`] = `"8/22/2017"`;

exports[`Formatting should format with default string 1`] = `"1/1/1970 12:00 am"`;

exports[`Formatting should format with default string 2`] = `"1/1/1970 12:00 am"`;

exports[`Formatting should format with default string 3`] = `"1/1/1970"`;

exports[`Formatting should format with default string 4`] = `"1/1/1970"`;

exports[`Formatting should format with defaults 1`] = `"1/1/1970"`;

exports[`Formatting should format with defaults 2`] = `"1/1/1970"`;

exports[`Formatting should format with individual options 1`] = `"1/1/1970"`;

exports[`Formatting should format with individual options 2`] = `"Jan 1, 1970"`;

exports[`Formatting should format with individual options 3`] = `"January 1, 1970"`;

exports[`Formatting should format with individual options 4`] = `"1/1/1970 12:00 am"`;

exports[`Formatting should format with individual options 5`] = `"1/1/1970"`;

exports[`Formatting should format with individual options 6`] = `"1/1/1970"`;

exports[`Formatting should format with individual options 7`] = `"12:00 am"`;

exports[`Formatting should format with individual options 8`] = `""`;

exports[`Formatting should format with options 1`] = `"1/1/1970 12:00 am"`;

exports[`Formatting should format with options 2`] = `"Jan 1, 1970 12:00 am"`;

exports[`Formatting should format with options 3`] = `"January 1, 1970 12:00 am"`;

exports[`Formatting should use a custom date format 1`] = `"January_1_1970_12_00_am"`;
